### PR TITLE
JCLOUDS-853: Improve Create VirtualMachine Deployment

### PR DESCRIPTION
**Pull request by hsbhathiya**
Conflicts:
	azurecompute/src/test/java/org/jclouds/azurecompute/features/VirtualMachineApiLiveTest.java

Fix Checkstyle Violations

Address live test failiures

Fix Checkstyle Violations

minor changes:

fix live and mocktests

minor changes

Code style changes

### CODE REVIEW

**Comment by hsbhathiya** [comment]
@nacx 
can you review this? . I am working on merging https://github.com/jclouds/jclouds-labs/pull/157 now.

File: azurecompute/src/main/java/org/jclouds/azurecompute/compute/AzureComputeServiceAdapter.java [hunk 1]
```diff
@@ -127,20 +131,61 @@
...
+               .build();
+
+         params = DeploymentParams.builder()
+               .name(name).roleParam(roleParam)
+               .virtualNetworkName(virtualNetworkName)
+               .externalEndpoints(externalEndpoints)
+               .subnetName(subnetName)
+               .reservedIpName(reservedIPAddress)
+               .build();
+      }
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/compute/AzureComputeServiceAdapter.java]: All code here is duplicated. Extract it all out of the conditionals and put only the configuration set in it. You're using builders, so there's no problem on doing that.

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/DataVirtualHardDiskParam.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+   public Builder toBuilder() {
+      return builder().fromDataVirtualHardDiskParams(this);
+   }
+
+   public static Builder builder() {
+      return new Builder();
+   }
+
+   public static final class Builder {
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/DataVirtualHardDiskParam.java]: The google auto version we use now supports builders. We should try to move to using them.

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/DataVirtualHardDiskParam.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+               .logicalDiskSizeInGB(in.logicalDiskSizeInGB()).mediaLink(in.mediaLink())
+               .sourceMediaLink(in.sourceMediaLink());
+      }
+
+   }
+
+   public static DataVirtualHardDiskParam create(String hostCaching, String diskLabel, String diskName, Integer lun,
+         Integer logicalDiskSizeInGB, URI mediaLink, URI sourceMediaLink) {
+      return new AutoValue_DataVirtualHardDiskParam(hostCaching, diskLabel, diskName, lun, logicalDiskSizeInGB,
+            mediaLink, sourceMediaLink);
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/DataVirtualHardDiskParam.java]: All values are nullable? When is it possible to have such an object with no values?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/DeploymentParams.java [hunk 4]
```diff
@@ -54,79 +49,33 @@
...
-    * {@link OSImage#name() name} of the user or platform image.
-    */
-   public abstract String sourceImageName();
-
-   /**
-    * Indicates the {@link OSImage#mediaLink() location} when {@link #sourceImageName() source} is a platform image.
-    */
-   public abstract URI mediaLink();
+   // TODO move to RoleParams/NetworkConfigurationParams
+   @Nullable public abstract List<ExternalEndpoint> externalEndpoints();
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/DeploymentParams.java]: This wasn't nullable before. Has this really changed?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/DeploymentParams.java [hunk 5]
```diff
@@ -54,79 +49,33 @@
...
-    * The ReservedIPName element is only available using version 2014-05-01 or higher.
-    *
-    * @return reserved IP.
-    */
-   @SinceApiVersion("2014-05-01")
-   @Nullable
-   public abstract String reservedIPName();
-
-   public abstract List<String> subnetNames();
+   @Nullable public abstract String reservedIpName();
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/DeploymentParams.java]: Why not keeping the `@SinceApiVersion` annotation?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/LinuxConfigurationSetParams.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+      public abstract String path();
+
+      public static PublicKey create(String fingerPrint, String path) {
+         return new AutoValue_LinuxConfigurationSetParams_PublicKey(fingerPrint, path);
+      }
+   }
+
+   @AutoValue
+   public abstract static class KeyPair {
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/LinuxConfigurationSetParams.java]: Why these two different classes to hold the same information?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/LinuxConfigurationSetParams.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+   public abstract String configurationSetType();
+
+   @Nullable public abstract String hostName();
+
+   public abstract String userName();
+
+   public abstract String userPassword();
+
+   public abstract Boolean disableSshPasswordAuthentication();
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/LinuxConfigurationSetParams.java]: Use the primitive type for non-nullable values.

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/OSVirtualHardDiskParam.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                     resizedSizeInGB);
+      }
+
+   }
+
+   public static OSVirtualHardDiskParam create(String hostCaching, String diskLabel, String diskName, URI mediaLink,
+         String sourceImageName, OSImage.Type os, URI remoteSourceImageLink, Integer resizedSizeInGB) {
+      return new AutoValue_OSVirtualHardDiskParam(hostCaching, diskLabel, diskName, mediaLink, sourceImageName, os,
+            remoteSourceImageLink, resizedSizeInGB);
+   }
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/OSVirtualHardDiskParam.java]: Same comment as before. All nullables?

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/WindowsConfigurationSetParams.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+      public static Credential create(String domain, String userName, String password) {
+         return new AutoValue_WindowsConfigurationSetParams_Credential(domain, userName, password);
+      }
+   }
+
+   @AutoValue
+   public abstract static class DomainJoin {
+
+      @Nullable public abstract List<Credential> credentials();
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/WindowsConfigurationSetParams.java]: We should avoid returning null collections. Prefer returning empty ones.

File: azurecompute/src/main/java/org/jclouds/azurecompute/domain/WindowsConfigurationSetParams.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+   @Nullable public abstract String computerName();
+
+   @Nullable public abstract String adminPassword();
+
+   @Nullable public abstract DomainJoin domainJoin();
+
+   @Nullable public abstract String adminUserName();
+
+   @Nullable public abstract List<CertificateSetting> storedCertificateSettings();
```
Comment (nacx) [inline azurecompute/src/main/java/org/jclouds/azurecompute/domain/WindowsConfigurationSetParams.java]: Same comment as before. Avoid null collections when possible.

**Comment by andrewgaul** [comment]
Closing this given the work in #279.

**Pull request closed without merging**